include/beman/net/detail/uring_context.hpp

```diff
@@ -0,0 +1,321 @@
+            error = ::std::error_code(errno, ::std::system_category());
+        }
+    }
+
+    auto native_handle(socket_id id) -> native_handle_type override { return sockets[id]; }
+
+    auto set_option(socket_id id, int level, int name, const void* data, ::socklen_t size, ::std::error_code& error)
+        -> void override {
+        if (::setsockopt(native_handle(id), level, name, data, size) < 0) {
+            error = ::std::error_code(errno, ::std::system_category());
+        }
+    }
+
+    auto bind(socket_id id, const endpoint& ep, ::std::error_code& error) -> void override {
+        if (::bind(native_handle(id), ep.data(), ep.size()) < 0) {
+            error = ::std::error_code(errno, ::std::system_category());
+        }
+    }
+
+    auto listen(socket_id id, int no, ::std::error_code& error) -> void override {
+        if (::listen(native_handle(id), no) < 0) {
+            error = ::std::error_code(errno, ::std::system_category());
+        }
+    }
+
+    auto submit() -> void {
+        int r = ::io_uring_submit(&ring);
+        if (r < 0) {
+            throw ::std::system_error(-r, ::std::system_category(), "io_uring_submit failed");
+        }
+        assert(submitting >= r);
+        submitting -= r;
+        outstanding += r;
+    }
+
+    auto get_sqe(io_base* completion) -> ::io_uring_sqe* {
+        auto sqe = ::io_uring_get_sqe(&ring);
+        while (sqe == nullptr) {
+            // if the submission queue is full, flush and try again
+            submit();
+            sqe = ::io_uring_get_sqe(&ring);
+        }
+        ::io_uring_sqe_set_data(sqe, completion);
+        ++submitting;
+        return sqe;
+    }
+
+    auto wait() -> ::std::tuple<int, io_base*> {
+        ::io_uring_cqe* cqe = nullptr;
+        int             r   = ::io_uring_wait_cqe(&ring, &cqe);
+        if (r < 0) {
+            throw ::std::system_error(-r, ::std::system_category(), "io_uring_wait_cqe failed");
+        }
+
+        assert(outstanding > 0);
+        --outstanding;
+
+        const int  res        = cqe->res;
+        const auto completion = ::io_uring_cqe_get_data(cqe);
+        ::io_uring_cqe_seen(&ring, cqe);
+
+        return {res, static_cast<io_base*>(completion)};
+    }
+
+    auto run_one() -> ::std::size_t override {
+        if (auto count = process_task(); count) {
+            return count;
+        }
+
+        if (submitting) {
+            // if we have anything to submit, batch the submit and wait in a
+            // single system call. this allows io_uring_wait_cqe() below to be
+            // served directly from memory
+            unsigned wait_nr = 1;
+            int      r       = ::io_uring_submit_and_wait(&ring, wait_nr);
+            if (r < 0) {
+                throw ::std::system_error(-r, ::std::system_category(), "io_uring_submit_and_wait failed");
+            }
+            assert(submitting >= r);
+            submitting -= r;
+            outstanding += r;
```

Comment:
As mentioned previously, I think I wouldn't tie `submitting` and `outstanding`: I don't see a reason why `outstanding` can't just track how much outstanding work there is, independent of whether it was submitted.